File: backend/src/services/enum.service.js
class EnumService {
  constructor(enumConstantsModel) {
    this.model = enumConstantsModel;
  }

  async listTypes() {
    const rows = await this.model.findAll({
      attributes: ["type"],
      group: ["type"],
      order: [["type", "ASC"]],
      raw: true,
    });
    return rows.map((r) => r.type);
  }

  async listByType(type, { page = 1, limit = 100 } = {}) {
    const offset = (page - 1) * limit;
    const { count, rows } = await this.model.findAndCountAll({
      where: { type },
      order: [["id", "ASC"]],
      limit,
      offset,
    });
    return { rows, total: count, page, limit };
  }

  async addValue(type, enumValue) {
    if (!type || String(type).trim() === "") {
      const err = new Error("type is required");
      err.status = 422;
      throw err;
    }
    if (enumValue === undefined || enumValue === null) {
      const err = new Error("enum value is required");
      err.status = 422;
      throw err;
    }

    const existing = await this.model.findOne({
      where: { type, enum: enumValue },
    });
    if (existing) {
      const err = new Error("Enum value already exists for this type");
      err.status = 409;
      throw err;
    }

    return this.model.create({ type, enum: String(enumValue) });
  }

  async deleteValue(id) {
    const row = await this.model.findByPk(id);
    if (!row) {
      const err = new Error("Enum value not found");
      err.status = 404;
      throw err;
    }
    await row.destroy();
  }
}

module.exports = EnumService;

File: backend-go/internal/service/enum_service.go
package service

import "gorm.io/gorm"

type EnumService struct {
	db *gorm.DB
}

func NewEnumService(db *gorm.DB) *EnumService {
	return &EnumService{db: db}
}

func (s *EnumService) ListTypes() ([]string, error) {
	var types []string
	err := s.db.Table("enum_constants").Distinct("type").Order("type ASC").Pluck("type", &types).Error
	return types, err
}

func (s *EnumService) ListByType(t string, page int, limit int) (map[string]any, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 {
		limit = 100
	}
	if limit > 500 {
		limit = 500
	}
	db := s.db.Table("enum_constants").Where("type = ?", t)
	var total int64
	if err := db.Count(&total).Error; err != nil {
		return nil, err
	}
	var rows []map[string]any
	if err := db.Order("id ASC").Limit(limit).Offset((page - 1) * limit).Find(&rows).Error; err != nil {
		return nil, err
	}
	return map[string]any{"rows": rows, "total": total, "page": page, "limit": limit}, nil
}

func (s *EnumService) AddValue(t string, value string) (map[string]any, error) {
	if t == "" {
		return nil, NewHttpError(422, "type is required")
	}
	if value == "" {
		return nil, NewHttpError(422, "enum value is required")
	}
	var count int64
	s.db.Table("enum_constants").Where("type = ? AND enum = ?", t, value).Count(&count)
	if count > 0 {
		return nil, NewHttpError(409, "Enum value already exists for this type")
	}
	row := map[string]any{"type": t, "enum": value}
	if err := s.db.Table("enum_constants").Create(&row).Error; err != nil {
		return nil, err
	}
	return row, nil
}

func (s *EnumService) DeleteValue(id string) error {
	res := s.db.Table("enum_constants").Where("id = ?", id).Delete(nil)
	if res.Error != nil {
		return res.Error
	}
	if res.RowsAffected == 0 {
		return NewHttpError(404, "Enum value not found")
	}
	return nil
}
